pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -209,34 +209,109 @@ func prepateMigrationFlags(isBlockMigration bool) libvirt.DomainMigrateFlags {
 
 }
 
+func gatherSharedVolumeForMigration(vmi *v1.VirtualMachineInstance) map[string]bool {
+	// This method collects all VMI volumes that should not be copied during
+	// live migration. Persistent volume claims without ReadWriteMany access mode
+	// should be filtered out earlier in the process
+	sharedVols := make(map[string]bool)
+	for _, volume := range vmi.Spec.Volumes {
+		volSrc := volume.VolumeSource
+		if volSrc.PersistentVolumeClaim != nil ||
+			(volSrc.HostDisk != nil && *volSrc.HostDisk.Shared) {
+			sharedVols[volume.Name] = true
+		}
+	}
+	return sharedVols
+}
+
+func getAllDomainDisks(dom cli.VirDomain) ([]api.Disk, error) {
+	xmlstr, err := dom.GetXMLDesc(0)
+	if err != nil {
+		return nil, err
+	}
+
+	var newSpec api.DomainSpec
+	err = xml.Unmarshal([]byte(xmlstr), &newSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	return newSpec.Devices.Disks, nil
+}
+
+func getDiskTargetsForMigration(dom cli.VirDomain, vmi *v1.VirtualMachineInstance) []string {
+	// This method collects all VMI disks that needs to be copied during live migration
+	// and returns a list of its target device names.
+	// Shared volues are being excluded.
+	copyDisks := []string{}
+	sharedVols := gatherSharedVolumeForMigration(vmi)
+
+	disks, err := getAllDomainDisks(dom)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("failed to parse domain XML to get disks.")
+	}
+	// the name of the volume should match the alias
+	for _, disk := range disks {
+		_, isSharedVolumeDisk := sharedVols[disk.Alias.Name]
+		if disk.ReadOnly != nil {
```

Comment:
so cloud-init/configmap/secret disks will get copied if they aren't marked as ReadOnly?  Maybe we want to always consider migrating those ephemeral files over so we can be certain they are the binary equivalent to what's on the source.

Reply:
> so cloud-init/configmap/secret disks will get copied if they aren't marked as ReadOnly?

Yes.

> Maybe we want to always consider migrating those ephemeral files over so we can be certain they are the binary equivalent to what's on the source.

Yes, It'll be better to just always copy it. I was worried that it'll be unexpected for the user. We will allow him to switch to virtio and set the disk to ReadOnly (none of it is a default) and then we will still copy the disks. Maybe we should just block setting these disks to readonly?
If you think that it's better to just copy it regardless, I can make it work here, it's not difficult.

Reply:
> We will allow him to switch to virtio and set the disk to ReadOnly (none of it is a default) and then we will still copy the disks. Maybe we should just block setting these disks to readonly?
If you think that it's better to just copy it regardless, I can make it work here, it's not difficult.

Yeah, if we dynamically create something during the launch flow like the cloud-init/secret/configmap disks, then I think we'll always need to copy those disks from the source to the target. Otherwise we can't guarantee they are binary equivalent. 

The different behavior based on the 'ReadOnly' flag here makes for a difficult to understand user story now that we're considering exceptions though. What are trying to optimize here for? If we're primarily just concerned about not copying ReadOnly ContainerDisks, maybe that should be what we key off of.

Reply:
Yeah, you're right. I've updated it to alway to copy the generated disks.